cpp/include/cuspatial/experimental/detail/pairwise_point_in_polygon.cuh

```diff
@@ -0,0 +1,139 @@
+                                                     poly_end,
+                                                     ring_offsets_first,
+                                                     num_rings,
+                                                     poly_points_first,
+                                                     num_poly_points);
+    result[idx]                = point_is_within;
+  }
+}
+
+}  // namespace detail
+
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OutputIt>
+OutputIt pairwise_point_in_polygon(Cart2dItA test_points_first,
+                                   Cart2dItA test_points_last,
+                                   OffsetIteratorA polygon_offsets_first,
+                                   OffsetIteratorA polygon_offsets_last,
+                                   OffsetIteratorB poly_ring_offsets_first,
+                                   OffsetIteratorB poly_ring_offsets_last,
+                                   Cart2dItB polygon_points_first,
+                                   Cart2dItB polygon_points_last,
+                                   OutputIt output,
+                                   rmm::cuda_stream_view stream)
+{
+  using T = iterator_vec_base_type<Cart2dItA>;
+
+  auto const num_test_points = std::distance(test_points_first, test_points_last);
+  auto const num_polys       = std::distance(polygon_offsets_first, polygon_offsets_last);
+  auto const num_rings       = std::distance(poly_ring_offsets_first, poly_ring_offsets_last);
+  auto const num_poly_points = std::distance(polygon_points_first, polygon_points_last);
+
+  static_assert(is_same_floating_point<T, iterator_vec_base_type<Cart2dItB>>(),
+                "Underlying type of Cart2dItA and Cart2dItB must be the same floating point type");
+  static_assert(
+    is_same<vec_2d<T>, iterator_value_type<Cart2dItA>, iterator_value_type<Cart2dItB>>(),
+    "Inputs must be cuspatial::vec_2d");
+
+  static_assert(cuspatial::is_integral<iterator_value_type<OffsetIteratorA>,
+                                       iterator_value_type<OffsetIteratorB>>(),
+                "OffsetIterators must point to integral type.");
+
+  static_assert(std::is_same_v<iterator_value_type<OutputIt>, int32_t>,
+                "OutputIt must point to 32 bit integer type.");
+
+  CUSPATIAL_EXPECTS(num_rings >= num_polys, "Each polygon must have at least one ring");
+  CUSPATIAL_EXPECTS(num_poly_points >= num_polys * 4, "Each ring must have at least four vertices");
+
+  CUSPATIAL_EXPECTS(num_test_points == num_polys,
+                    "Must pass in an equal number of points and polygons");
+
+  // TODO: introduce a validation function that checks the rings of the polygon are
+  // actually closed. (i.e. the first and last vertices are the same)
+
+  auto constexpr block_size = 256;
+  auto const num_blocks     = (num_test_points + block_size - 1) / block_size;
```

Comment:
You can now use grid_1d in cuspatial!

```suggestion
  auto [threads_per_block, num_blocks] = grid_id(num_test_points);
```

You should `#include <cuspatial/cuda_utils.cuh>`

Reply:
Done.